backstory/src/main/java/com/google/sps/story/data/NLServiceClient.java

```diff
@@ -0,0 +1,185 @@
+
+  /**
+   * Takes a list of words and sorts them by word type.
+   * Returns a map which has word types as keys and then words
+   * in a list linked to their relevant word type (e.g. "Mary" would
+   * be sorted into the PROPER_NOUN list and "calm" would be sorted
+   * into the adjective list).
+   *
+   * @param words the words to sort by word type
+   * @return a map with lists of words sorted by their word type and with their
+   *    word type as a key
+   */
+  public Map<WordType, List<String>> sortByWordType(List<String> words) {
+    Map<WordType, List<String>> map = new HashMap<WordType, List<String>>();
+
+    List<String> singleWords = new ArrayList<String>();
+
+    // remove multi-word inputs first and store as multiword nouns
+    for (String text : words) {
+      if (text.contains(" ")) {
+        if (map.containsKey(WordType.MULTIWORD_NOUN)) {
+          List<String> wordList = map.get(WordType.MULTIWORD_NOUN);
+          wordList.add(text);
+        } else {
+          ArrayList<String> wordList = new ArrayList<String>();
+          wordList.add(text);
+          map.put(WordType.MULTIWORD_NOUN, wordList);
+        }
+      } else {
+        singleWords.add(text);
+      }
+    }
+
+    for (String word : singleWords) {
+      Document doc = buildDocumentFromText(word);
+
+      AnalyzeSyntaxResponse response = client.analyzeSyntax(doc);
+
+      Token token = response.getTokens(0);
+      PartOfSpeech partOfSpeech = token.getPartOfSpeech();
+
+      WordType type = WordType.UNUSABLE;
+
+      switch (partOfSpeech.getTag()) {
+        case VERB:
+          if (isGerund(word)) {
+            type = WordType.GERUND;
+          }
+          break;
+        case NOUN:
+          if (isGerund(word)) {
+            type = WordType.GERUND;
+            break;
+          }
+
+          if (partOfSpeech.getProper() == Proper.PROPER) {
+            type = WordType.PROPER_NOUN;
+          } else {
+            type = WordType.NOUN;
+          }
+          break;
+
+        case ADJ:
+          type = WordType.ADJECTIVE;
+          break;
+      }
+
+      if (map.containsKey(type)) {
+        List<String> wordList = map.get(type);
+        wordList.add(word);
+      } else {
+        ArrayList<String> wordList = new ArrayList<String>();
+        wordList.add(word);
+        map.put(type, wordList);
+      }
+    }
+
+    return map;
+  }
+
+  /**
+   * Build and return a document from the given test.
+   *
+   * @param text the text to build a document from
+   * @return a Document that holds the current text
+   */
+  private Document buildDocumentFromText(String text) {
+    Document doc = Document.newBuilder().setContent(text).setType(Type.PLAIN_TEXT).build();
+
+    return doc;
+  }
+
+  /**
+   * Checks if a word is a gerund using heuristics (ending in "ing) and NL API
+   *
+   * @param word the word to check if it's a gerund
+   * @return true, if it's a gerund; false, otherwise
+   */
+  private boolean isGerund(String word) {
```

Comment:
Nit: I like how you check heuristics first... it may help to highlight in the docs that this may make an API call, so the user knows that this is a (relatively) expensive method call.

Reply:
I will note that. Thanks!

Reply:
Done